Add wheel info subcommand

killerdevildog opened the pull request:
Implements GitHub issue #639: Add wheel info subcommand to display metadata about wheel files without unpacking them.

Features:
- Shows package name, version, and build information
- Displays wheel format version and generator
- Lists supported Python versions, ABI, and platform tags
- Shows package metadata (summary, author, license, classifiers)
- Displays dependencies and file count/size information
- Optional verbose mode with detailed file listing
- Comprehensive error handling for missing files

Changes:
- Add src/wheel/_commands/info.py with main implementation
- Update src/wheel/_commands/__init__.py to register new command
- Add tests/commands/test_info.py with comprehensive test coverage
- Add docs/reference/wheel_info.rst with usage documentation
- Update docs/reference/index.rst to include new command docs
- Update docs/user_guide.rst with info command examples
- Update docs/manpages/wheel.rst with info command reference

Addresses #639
See #639

killerdevildog committed:
Add wheel info subcommand

Implements GitHub issue #639: Add wheel info subcommand to display
metadata about wheel files without unpacking them.

Features:
- Shows package name, version, and build information
- Displays wheel format version and generator
- Lists supported Python versions, ABI, and platform tags
- Shows package metadata (summary, author, license, classifiers)
- Displays dependencies and file count/size information
- Optional verbose mode with detailed file listing
- Comprehensive error handling for missing files

Changes:
- Add src/wheel/_commands/info.py with main implementation
- Update src/wheel/_commands/__init__.py to register new command
- Add tests/commands/test_info.py with comprehensive test coverage
- Add docs/reference/wheel_info.rst with usage documentation
- Update docs/reference/index.rst to include new command docs
- Update docs/user_guide.rst with info command examples
- Update docs/manpages/wheel.rst with info command reference

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

agronholm commented:
Have you faced such criticism in this project, or another PyPA project, or from me? What matters to me is that you can explain what your code does and why. I have rejected AI generated patches in the past because the authors could not explain them to me, and the quality of AI code generated as of today tends to be subpar. But I certainly don't follow people around harassing them for this.

killerdevildog commented:
Thanks for clarifying. The issues were in other PyPA projects, not this one or from you. I'm able to explain the code fully—it's my work with some LLM aid for initial drafting, but I understand and tested it thoroughly. Happy to address any questions.

The PR directly implements the feature requested in Issue #639, which asks for a new wheel info subcommand to display metadata (e.g., package name, version, dependencies) for a .whl file or PyPI package without unpacking it. The changes add a new command module, tests, and documentation to achieve this, reading data straight from the wheel's zip structure for efficiency.

agronholm commented:
Thanks for the explanation. I'm not currently working on wheel, and I had to yank the last couple releases which makes me apprehensive about any changes I make to the repo. This PR, however, doesn't have such issues as it's completely new functionality, so I will get back to this once I resume work here.

agronholm committed:
Merge branch 'main' into feature/wheel-info-subcommand

agronholm reviewed (changes requested):
A changelog entry needs to be added in `docs/news.rst` and the issues I pointed out need to be resolved.

killerdevildog committed:
Address maintainer review feedback on wheel info subcommand

- Remove unused tmp_path_factory parameter from test_info_basic and test_info_verbose
- Use pytest.raises(FileNotFoundError, match=...) instead of try/except in test_info_nonexistent_file
- Fix test_info_help to not depend on argv[0] being 'wheel' (fixes Python 3.14 CI failure)
- Add changelog entry in docs/news.rst for the new wheel info subcommand

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

killerdevildog commented:
Hi @agronholm, thanks for the review feedback. I've pushed a commit addressing all your requested changes:

Removed unused tmp_path_factory — Removed the unnecessary TempPathFactory parameter from test_info_basic() and test_info_verbose().

Replaced try/except with pytest.raises — test_info_nonexistent_file now uses pytest.raises(FileNotFoundError, match="Wheel file not found: nonexistent.whl") to properly test the error case, calling info() directly since the CLI runner catches WheelError internally.

Fixed the Python 3.14 test failure — test_info_help was asserting "usage: wheel info" but sys.argv[0] differs depending on how the tests are invoked (e.g. python -m pytest instead of wheel). Updated the assertion to check for the help content without depending on the program name.

Added changelog entry — Added an entry under UNRELEASED in news.rst referencing issue #639.

All tests pass on both Python 3.12 and 3.14.

killerdevildog committed:
Remove unused variables flagged by ruff (F841)

Remove python_tag, abi_tag, and platform_tag assignments that were
extracted from the wheel filename but never used.

agronholm reviewed (approved):
I'll solicit outside opinions on the output of the command before merging.

pfmoore commented on src/wheel/_commands/info.py:
```diff
@@ -0,0 +1,124 @@
+                    for tag in sorted(tags):  # Sort tags for consistent output
+                        print(f"  {tag}")
+
+                generator = wheel_metadata.get("Generator")
```
For future proofing, this should probably allow for multiple `Generator` values. See https://discuss.python.org/t/generator-metadata-of-wheels-created-using-multiple-tools/106016 for context.

agronholm commented:
The repeated-lines approach is what I would have assumed as well, so go with that. As for the JSON output, that can be deferred to a new PR.

killerdevildog committed:
Support multiple Generator values in wheel info output

killerdevildog committed:
Improve tests and fix changelog placement for wheel info

- Move changelog entry to UNRELEASED section per project release process
  (was incorrectly placed under the already-released 0.46.2 heading)
- Add test for zero Generator values (test_info_no_generator)
- Add exact count assertion for multiple Generator lines
- Refactor test helpers: extract _build_wheel_with_modified_metadata()
  and _capture_info_output() for reusable wheel modification with
  proper RECORD hash updates
- Move all imports to module level (no more inline imports in tests)
- 8 tests now cover: basic info, single generator, multiple generators,
  zero generators, verbose output, nonexistent file, help, and -v flag

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

agronholm committed:
Redirect warnings to stderr in info.py

Updated warning messages to print to stderr.

killerdevildog commented:
@agronholm I will work on that this week. Thanks for letting me participate.
